44050 new core metrics

sgsteph opened the pull request:
Changes needed to moved to Amplitude metrics

josh-t commented on hooks/log_metrics.py:
```diff
@@ -11,12 +11,18 @@
+        .. warning::
+            This method is deprecated is not executed anymore.
+
+            Please check :meth:`LogMetrics.execute2`.
```
change the name here as well. :)

josh-t commented on python/tank/platform/engine.py:
```diff
@@ -661,6 +650,36 @@ def created_qt_dialogs(self):
+        }
+
+    @property
+    def host_info_string(self):
```
do we need this?

josh-t commented on python/tank/util/metrics.py:
```diff
@@ -42,6 +46,12 @@ class MetricsQueueSingleton(object):
 
     """
 
+    MAXIMUM_QUEUE_SIZE = 20
```
this seems low to me. any reason for 20 specifically? can we bump it a bit to maybe 100 or so?

nmontmarquette replied:
The value of 20 is totally arbitrary, I'm actually missing a TODO: so we discuss on the maximum size.
 
I agree that 100 seems like a fair size, although if it was ever to queue up to 100 metric event, we probably want to ensure that it doesn't get sent in a single large payload. The server won't accept more than 10 at once. I would like to add a test to make sure it get broken in server-compatible payloads.

josh-t commented on python/tank/util/metrics.py:
```diff
@@ -62,19 +72,20 @@ def __new__(cls, *args, **kwargs):
 
             # The underlying collections.deque instance
-            metrics_queue._queue = deque()
+            metrics_queue._queue = deque(maxlen=MetricsQueueSingleton.MAXIMUM_QUEUE_SIZE)
```
this could be simplified to 

```python
 metrics_queue._queue = deque(maxlen=self.MAXIMUM_QUEUE_SIZE)
```

josh-t commented on python/tank/util/metrics.py:
```diff
@@ -309,31 +323,99 @@ def _dispatch(self, metrics):
-        self._engine.tank.execute_core_hook(
+        self._engine.tank.execute_core_hook_method(
             constants.TANK_LOG_METRICS_HOOK_NAME,
+            # 'execute' was used with previous metrics system which has
```
i don't think we need to include this comment.

josh-t commented on python/tank/util/metrics.py:
```diff
@@ -344,80 +426,59 @@ def data(self):
-    This method simply adds the metric to the dispatch queue.
+    """ 
+    This method is deprecated and shouldn't be used anymore.
+    Please use the `log_metric_event` method.
```
Should be `EventMetric.log`

josh-t commented on python/tank/util/metrics.py:
```diff
@@ -344,80 +426,59 @@ def data(self):
-        already been logged. Defaults to ``False``.
+    """ 
+    This method is deprecated and shouldn't be used anymore.
+    Please use the `log_metric_event` method.
```
same comment here.

josh-t commented on python/tank/util/metrics.py:
```diff
@@ -344,80 +426,59 @@ def data(self):
+def log_user_attribute_metric(attr_name, attr_value, log_once=False):
+    """ 
+    This method is deprecated and shouldn't be used anymore.
+    Please use the `log_metric_event` method.
```
same comment here.

josh-t commented on hooks/log_metrics.py:
```diff
@@ -11,12 +11,18 @@
-        """Called when Toolkit logs user metrics.
+        """
+        .. warning::
+            This method is deprecated is not executed anymore.
```
minor: Add `and` before `is`.

josh-t commented on python/tank/util/metrics.py:
```diff
@@ -344,80 +439,59 @@ def data(self):
-    def __repr__(self):
-        """Official str representation of the user attribute metric."""
-        return "%s.%s" % (self._data["attr_name"], self._data["attr_value"])
+        log.debug("EventMetric.log('%s', '%s')" % (
```
I suggest commenting this out. Devs can uncomment as needed, but likely not needed by clients? Also suggest moving the `import` for `LogManager` to here so that everything can be commented/uncommented at once.

nmontmarquette replied:
Actually logging is not that useful, I found out you can fully debug everything from this module by attaching a python debugger. You need to be running tk-desktop and be in a 'project'

nmontmarquette commented on python/tank/platform/engine.py:
```diff
@@ -267,8 +267,6 @@ def __init__(self, tk, context, engine_instance_name, env):
 
         self.log_debug("Init complete: %s" % self)
-        # Emit an event for the software being launched
-        self.log_metric("Launched Software")
```
@robblau Note that moving this to the `tk-multi-launchapp` we lose the ability to log Host App and Host App Version from this event ( since engine are not yet started in `tk-multi-launchapp` )

sgsteph commented:
@josh-t The code is ready for another review pass:
- Changed the code to make clear that "Toolkit" group is for internal use.
- Removed pre-defined groups which shouldn't be used from Toolkit.
- Fire the `log_metric` hook, even if the SG endpoint is not available, added unit test for it.
- Some code/docstring cleanup.
- Removed a couple of `log_metric` calls which are deprecated.

josh-t commented on python/tank/util/metrics.py:
```diff
@@ -204,25 +212,40 @@ def workers(self):
+    """
+
+    # List of Event names suported by our backend
+    SUPPORTED_EVENTS = [
```
minor: suggest formatting these as one event string per line. that'll make it clearer in the history when we add/remove individual events.

josh-t commented on python/tank/util/metrics.py:
```diff
@@ -247,44 +270,95 @@ def __init__(self, engine):
+            # Only send internal Toolkit events
+            if metric.is_internal_event:
+                data = metric.data
+                if data["event_name"] not in self.SUPPORTED_EVENTS:
```
perhaps we should store the `event_name` as a property so that we don't loose that information? might be useful to evaluate what the "unknown" events are.

robblau replied:
From the Taxonomy spreadsheet there should be two properties: event name, and event data.  That way we can mine it to see what's getting captured and figure out if we need new official events or some ignore logic.

sgsteph replied:
Sorry, totally missed that...

sgsteph commented on tests/util_tests/test_metrics.py:
```diff
@@ -8,62 +10,722 @@
+        self.assertTrue(isinstance(server_received_metric["event_properties"][EventMetric.KEY_HOST_APP], unicode))
+        self.assertTrue(isinstance(server_received_metric["event_properties"][EventMetric.KEY_HOST_APP_VERSION], unicode))
+        self.assertTrue(isinstance(server_received_metric["event_properties"][EventMetric.KEY_APP], unicode))
+        self.assertTrue(isinstance(server_received_metric["event_properties"][EventMetric.KEY_APP_VERSION], unicode))
```
Ignoring, false positive.

sgsteph commented on tests/util_tests/test_metrics.py:
```diff
@@ -8,62 +10,722 @@
+
+        self.assertTrue(isinstance(server_received_metric["event_properties"][EventMetric.KEY_HOST_APP], unicode))
+        self.assertTrue(isinstance(server_received_metric["event_properties"][EventMetric.KEY_HOST_APP_VERSION], unicode))
+        self.assertTrue(isinstance(server_received_metric["event_properties"][EventMetric.KEY_APP], unicode))
```
Ignoring, false positive.

sgsteph commented on tests/util_tests/test_metrics.py:
```diff
@@ -8,62 +10,722 @@
+        self.assertTrue(isinstance(server_received_metric["event_properties"], dict))
+
+        self.assertTrue(isinstance(server_received_metric["event_properties"][EventMetric.KEY_HOST_APP], unicode))
+        self.assertTrue(isinstance(server_received_metric["event_properties"][EventMetric.KEY_HOST_APP_VERSION], unicode))
```
Ignoring, false positive.

sgsteph commented on tests/util_tests/test_metrics.py:
```diff
@@ -8,62 +10,722 @@
+        self.assertTrue(isinstance(server_received_metric["event_name"], unicode))
+        self.assertTrue(isinstance(server_received_metric["event_properties"], dict))
+
+        self.assertTrue(isinstance(server_received_metric["event_properties"][EventMetric.KEY_HOST_APP], unicode))
```
Ignoring, false positive.

sgsteph commented on tests/util_tests/test_metrics.py:
```diff
@@ -8,62 +10,722 @@
+        self.assertTrue("ListProp" in preserved_properties)
+
+        self.assertTrue(isinstance(server_received_metric["event_group"], unicode))
+        self.assertTrue(isinstance(server_received_metric["event_name"], unicode))
```
Ignoring, false positive.

sgsteph commented on tests/util_tests/test_metrics.py:
```diff
@@ -8,62 +10,722 @@
+        self.assertTrue("DictProp" in preserved_properties)
+        self.assertTrue("ListProp" in preserved_properties)
+
+        self.assertTrue(isinstance(server_received_metric["event_group"], unicode))
```
Ignoring, false positive.

sgsteph commented on tests/util_tests/test_metrics.py:
```diff
@@ -8,62 +10,722 @@
+        self.assertTrue(isinstance(server_received_metric["event_properties"][EventMetric.KEY_HOST_APP], unicode))
+        self.assertTrue(isinstance(server_received_metric["event_properties"][EventMetric.KEY_HOST_APP_VERSION], unicode))
+        self.assertTrue(isinstance(server_received_metric["event_properties"][EventMetric.KEY_APP], unicode))
+        self.assertTrue(isinstance(server_received_metric["event_properties"][EventMetric.KEY_APP_VERSION], unicode))
```
Ignoring, false positive.

sgsteph commented on tests/util_tests/test_metrics.py:
```diff
@@ -8,62 +10,722 @@
+
+        self.assertTrue(isinstance(server_received_metric["event_properties"][EventMetric.KEY_HOST_APP], unicode))
+        self.assertTrue(isinstance(server_received_metric["event_properties"][EventMetric.KEY_HOST_APP_VERSION], unicode))
+        self.assertTrue(isinstance(server_received_metric["event_properties"][EventMetric.KEY_APP], unicode))
```
Ignoring, false positive.

sgsteph commented on tests/util_tests/test_metrics.py:
```diff
@@ -8,62 +10,722 @@
+        self.assertTrue(isinstance(server_received_metric["event_properties"], dict))
+
+        self.assertTrue(isinstance(server_received_metric["event_properties"][EventMetric.KEY_HOST_APP], unicode))
+        self.assertTrue(isinstance(server_received_metric["event_properties"][EventMetric.KEY_HOST_APP_VERSION], unicode))
```
Ignoring, false positive.

sgsteph commented on tests/util_tests/test_metrics.py:
```diff
@@ -8,62 +10,722 @@
+        self.assertTrue(isinstance(server_received_metric["event_name"], unicode))
+        self.assertTrue(isinstance(server_received_metric["event_properties"], dict))
+
+        self.assertTrue(isinstance(server_received_metric["event_properties"][EventMetric.KEY_HOST_APP], unicode))
```
Ignoring, false positive.

sgsteph commented on tests/util_tests/test_metrics.py:
```diff
@@ -8,62 +10,722 @@
+        self.assertTrue("ListProp" in server_received_metric["event_properties"])
+
+        self.assertTrue(isinstance(server_received_metric["event_group"], unicode))
+        self.assertTrue(isinstance(server_received_metric["event_name"], unicode))
```
Ignoring, false positive.

sgsteph commented on tests/util_tests/test_metrics.py:
```diff
@@ -8,62 +10,722 @@
+        self.assertTrue("DictProp" in server_received_metric["event_properties"])
+        self.assertTrue("ListProp" in server_received_metric["event_properties"])
+
+        self.assertTrue(isinstance(server_received_metric["event_group"], unicode))
```
Ignoring, false positive.
